### PR TITLE
S107 user timeline

### PR DESCRIPTION
**Pull request by HadarRose**
Added user timeline to page

### CODE REVIEW

File: src/js/components/timeline.js [hunk 1]
```diff
@@ -1,75 +1,33 @@
...
-            tweets: [], // tweets retrieved 
-            isLoaded: false, // flags if request was loaded or not
-            error: null, // error message
-        };
-        this.errorMessage = "Something went wrong, please contact systems administrator.";
-    }
-
-    componentDidMount() { // calls requestTimeline upon component being mounted
-        this.requestTimeline();
+        if(this.props?.type){
```
Comment (jcorteza) [inline src/js/components/timeline.js]: I've never seen this before, how does this work with the question mark there? Also, it's probably not necessary to attach the value of className the the instance of Timeline. You could always declare a local variable later in the render method

File: src/js/components/timeline.js [hunk 2]
```diff
@@ -1,75 +1,33 @@
...
+                    <h3>{this.props.title}</h3>
+                </div>
+                <button id="timeline-button" onClick={() => this.getTimeline()}>Refresh</button>
+                <TweetsContainer type={this.props?.type} ref={'timeline'}/>
             </div>
         );
     }
+}
+
+Timeline.defaultProps = {
```
Comment (jcorteza) [inline src/js/components/timeline.js:33]: This is cool, didn't know react had this feature

File: src/js/components/tweetBlock.js [hunk 3]
```diff
@@ -1,23 +1,27 @@
 import React from 'react';
 
 export default class TweetBlock extends React.Component {
+    constructor(props){
+        super(props);
+    }
+
     render(){
         let tweetLink = "https://twitter.com/" +
```
Comment (jcorteza) [inline src/js/components/tweetBlock.js]: a template literal would increase the readability of this string

File: src/js/components/tweetsContainer.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            isLoaded: false, // flags if request was loaded or not
+            error: null, // error message
+        };
+        let typeProperties = TYPES[this.props.type];
+
+        /* HANDLING TYPE AND PROPERTIES */
+        // choose service method
+        if(this.props?.serviceMethod){
+            this.serviceMethod = this.props.serviceMethod;
+        }
```
Comment (jcorteza) [inline src/js/components/tweetsContainer.js]: your bracket is not aligned with your `else if`

File: src/js/components/tweetsContainer.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            tweets: response.data,
+            error: null
+        });
+    } 
+
+    declareError(err){
+        console.error(err);
+        this.setState({
+            isLoaded: true,
+            tweet: null,
```
Comment (jcorteza) [inline src/js/components/tweetsContainer.js]: do you mean `tweets` and does your component work if tweet is null when the default value is an empty array?

File: src/js/components/tweetsContainer.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        if(this.state.isLoaded && this.state.error){ // if loaded and an error, render error message
+            content = (
+                <div className="error-message">
+                    {this.errorMessage}
+                </div>
+            );
+        } else if(this.state.isLoaded){ // if loaded and no error, render a list of TweetBlock
+            content = (
+                    this.state.tweets.map(t => (
+                        <TweetBlock tweet={t} additionalProperties={this.blockProps}/>
```
Comment (jcorteza) [inline src/js/components/tweetsContainer.js]: probably need to add keys to the list items

File: src/styles/index.scss [hunk 7]
```diff
@@ -1,95 +1,133 @@
...
-                flex: 0 0 15%; /*this sets user div's width to always be 15% regardless of flexing*/
-                > #user-icon {
-                    border-radius: 50%;
-                    height: 65px;
+            }
+
+            .tweets-container {
+                border: 1pt solid black;
+                margin: 5px 0px;
+                //flex: 0 0 40%;
```
Comment (jcorteza) [inline src/styles/index.scss]: please remove if not using this line

File: tests/timelinePage.test.js [hunk 8]
```diff
@@ -16,11 +16,7 @@ describe('Timeline Page', () =>{
         expect(wrapper.find('div.title-div').length).toEqual(1);
     });
 
-    it('should render a timeline button', () => {
-        expect(wrapper.find('button#timeline-button').length).toEqual(1); 
-    });
-
-    it('should render a Timeline component', () =>{
-        expect(wrapper.find('Timeline').length).toEqual(1);
+    it('should render two Timeline components', () =>{
```
Comment (jcorteza) [inline tests/timelinePage.test.js]: nit: please add space between your arrows and brackets

File: src/styles/index.scss [hunk 9]
```diff
@@ -1,95 +1,133 @@
...
-                    font-size: 75%;
+                    &:last-child{
+                        border-bottom-style: none;
+                    }
+        
+                    .user-div  {
+                        text-align: center;
+                        margin-right: 30px;
+                        flex: 0 0 15%; /*this sets user div's width to always be 15% regardless of flexing*/
+                        > #user-icon {
```
Comment (jcorteza) [inline src/styles/index.scss]: user-icon seems like something that would be reoccurring throughout your page/component. seems like you may have duplicate ids

File: src/styles/index.scss [hunk 10]
```diff
@@ -1,95 +1,133 @@
 $font-stack: 'Times New Roman', Times, serif;
 
+// margin handing variables
+$timeline-separating-margin: 20px;
```
Comment (jcorteza) [inline src/styles/index.scss:4]: like your use of scss variables!

**Review by jcorteza — CHANGES_REQUESTED** [review]

File: src/js/components/tweetsContainer.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        return {
+            default: () => promiseTimeline(),
+            self: () => promiseTimelineSelf().then(
+                (response) => {
+                    if(response?.data?.length == 0){ // if no tweets, set error message to special error message
+                        this.errorMessage = 'No tweets are available, post a tweet!';
+                        throw new Error('No tweets');
+                    } else {
+                        response.data.map((tweet) => {  // remove twitterHandle from tweets that are passed to TweetBlock
+                            delete tweet.user.twitterHandle;
```
Comment (kev-chan) [inline src/js/components/tweetsContainer.js]: why the deletion? If someone was to pass data in, you may mistakenly modify the incoming data.  Also  reusable components

File: src/js/components/tweetsContainer.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
+import React from 'react';
+import TweetBlock from './tweetBlock';
+import {promiseTimeline, promiseTimelineSelf} from '../services/timelineService';
+
+export const DEFAULT_ERROR_MESSAGE = 'Something went wrong, please contact systems administrator.';
+
+export default class TweetsContainer extends React.Component {
```
Comment (kev-chan) [inline src/js/components/tweetsContainer.js:7]: this is much more readable and easier to follow.

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/js/components/tweetsContainer.js (hunk 11 above)
Comment (HadarRose) [reply src/js/components/tweetsContainer.js]: I think I understand. Would defining a new variable instead work? For example: 
`let resp = {
                            data: response.data.map((tweet) => {  TweetBlock
                                delete tweet.user.twitterHandle;
                                return tweet;
                            })
                        }
                        return resp;`
Also, could you clarify what you meant by 'Also reusable components'?

Comment (kev-chan) [reply src/js/components/tweetsContainer.js]: yeah it's a difference between data vs component/rendering

Comment (HadarRose) [reply src/js/components/tweetsContainer.js]: Updated

File: src/js/components/tweetsContainer.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                        throw new Error('No tweets');
+                    } else {
+                        return response;
+                    }
+                }
+            )
+        };
+    }
+
+    get TYPE_BLOCK_PROPS(){
```
Comment (kev-chan) [inline src/js/components/tweetsContainer.js]: if you're trying to make it generic, don't have different blocks of information as it'll get messy quickly

File: src/js/components/tweetsContainer.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        if(this.state.isLoaded && this.state.error){ // if loaded and an error, render error message
+            content = (
+                <div className="error-message">
+                    {this.errorMessage}
+                </div>
+            );
+        } else if(this.state.isLoaded){ // if loaded and no error, render a list of TweetBlock
+            content = (
+                    this.state.tweets.map(t => (
+                        <TweetBlock tweet={t} key={t.id} hidden={this.blockProps}/>
```
Comment (kev-chan) [inline src/js/components/tweetsContainer.js]: hidden makes it sound like the entire component is hidden based on the prop name

File: src/js/components/tweetBlock.js [hunk 15]
```diff
@@ -1,23 +1,25 @@
...
             return (
             <div className="tweet-block">
                 <div className="user-div">
-                    <img src={this.props.tweet.user.profileImageUrl} id="user-icon"></img>
+                    <img src={this.props.tweet.user.profileImageUrl} className="user-icon"></img>
                     <div className="user-name">
                         {this.props.tweet.user.name}
                     </div>
-                    <div className="user-handle">
+                    <div className="user-handle" hidden={this.props?.hidden?.handle}>
```
Comment (kev-chan) [inline src/js/components/tweetBlock.js]: why the decision to just hide it?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/js/components/tweetBlock.js (hunk 15 above)
Comment (HadarRose) [reply src/js/components/tweetBlock.js]: React's conditional rendering makes the component less readable in my opinion (and also doesn't seem to work the way it is documented). I will give it another shot, but this seemed like a pretty good solution overall.

**Pull request merged**